gc/src/lib.rs

```diff
@@ -694,6 +779,51 @@ pub struct GcCellRefMut<'a, T: Trace + ?Sized + 'static> {
     value: &'a mut T,
 }
 
+impl<'a, T: Trace + ?Sized> GcCellRefMut<'a, T> {
+    /// Makes a new `GcCellRefMut` for a component of the borrowed data, e.g., an enum
+    /// variant.
+    ///
+    /// The `GcCellRefMut` is already mutably borrowed, so this cannot fail.
+    ///
+    /// This is an associated function that needs to be used as
+    /// `GcCellRefMut::map(...)`. A method would interfere with methods of the same
+    /// name on the contents of a `GcCell` used through `Deref`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use gc::{GcCell, GcCellRefMut};
+    ///
+    /// let c = GcCell::new((5, 'b'));
+    /// {
+    ///     let b1: GcCellRefMut<(u32, char)> = c.borrow_mut();
+    ///     let mut b2: GcCellRefMut<u32> = GcCellRefMut::map(b1, |t| &mut t.0);
+    ///     assert_eq!(*b2, 5);
+    ///     *b2 = 42;
+    /// }
+    /// assert_eq!(*c.borrow(), (42, 'b'));
+    /// ```
+    #[inline]
+    pub fn map<U, F>(orig: Self, f: F) -> GcCellRefMut<'a, U>
+    where
+        U: Trace + ?Sized,
+        F: FnOnce(&mut T) -> &mut U,
+    {
+        let value = unsafe { &mut *(orig.value as *mut T) };
+
+        let ret = GcCellRefMut {
+            flags: orig.flags,
+            value: f(value),
+        };
```

Comment:
It looks like this new `GcCellRefMut` unroots the wrong `value` when dropped, causing a panic (#117).